dialect/sql/builder.go

```diff
@@ -701,13 +726,88 @@ func (i *InsertBuilder) Query() (string, []interface{}) {
 			i.WriteByte('(').Args(v...).WriteByte(')')
 		}
 	}
+
+	if len(i.conflictColumns) > 0 {
+		// Update on conflict
+		i.buildConflictHandling()
+	}
+
 	if len(i.returning) > 0 && i.postgres() {
 		i.WriteString(" RETURNING ")
 		i.IdentComma(i.returning...)
 	}
 	return i.String(), i.args
 }
 
+func (i *InsertBuilder) buildConflictHandling() {
+	switch i.Dialect() {
+	case dialect.Postgres, dialect.SQLite:
+		i.Pad().
+			WriteString("ON CONFLICT").
+			Pad().
+			Nested(func(b *Builder) {
+				b.IdentComma(i.conflictColumns...)
+			}).
+			Pad()
+
+		switch i.onConflictOp {
+		case OpResolveWithNewValues:
+			i.WriteString("DO UPDATE SET").Pad()
+			for j, c := range i.columns {
+				if j > 0 {
+					i.Comma()
+				}
+				i.Ident(c).WriteOp(OpEQ).Ident("excluded").WriteByte('.').Ident(c)
+			}
+		case OpResolveWithIgnore:
+			i.WriteString("DO UPDATE SET").Pad()
+			for j, c := range i.columns {
+				if j > 0 {
+					i.Comma()
+				}
+				// Ignore conflict by setting column to itself e.g. "c" = "c"
+				i.Ident(c).WriteOp(OpEQ).Ident(c)
+			}
+		case OpResolveWithAlternateValues:
+			i.WriteString("DO UPDATE SET").Pad()
+			writeUpdateValues(i, i.updateColumns, i.updateValues)
+		}
+
+	case dialect.MySQL:
+		i.Pad().WriteString("ON DUPLICATE KEY UPDATE ")
+
+		switch i.onConflictOp {
+		case OpResolveWithIgnore:
+			for j, c := range i.columns {
+				if j > 0 {
+					i.Comma()
+				}
+				// Ignore conflict by setting column to itself e.g. `c` = `c`
+				i.Ident(c).WriteOp(OpEQ).Ident(c)
+			}
```

Comment:
Move this to a helper function `writeIgnoreValues` or something else, and use it here and in the block above.